### PR TITLE
Git clone shallow to improve new importer instance performance

### LINKED ISSUE #576: Improve importer performance.

### PR DESCRIPTION
**Pull request by another-rex**
A shallow clone that only clones commits up to the previous synced date. This avoids having to clone the entire repository history when unneeded. A pull/fetch will only fetch future commits, without needing to fill up the entire history. 

A potential issue is if default branch changes to a completely different branch (not just renamed), this will mean potentially needing more history than what is available since the last sync time. This is very rare and I am not aware of any reason for doing this.

Fixes #576

### CODE REVIEW

File: osv/repos.py [hunk 1]
```diff
@@ -81,7 +81,7 @@ def _checkout_branch(repo, branch):
   repo.reset(remote_branch.target, pygit2.GIT_RESET_HARD)
 
 
-def clone(git_url, checkout_dir, git_callbacks=None):
+def clone(git_url, checkout_dir, git_callbacks=None, last_update_date=None):
```
Comment (andrewpollock) [inline osv/repos.py:84]: Could you:

at a minimum, add a docstring that describes the parameters, including the expected format of this date
preferably, add some input validation for `last_update_date`

Comment (another-rex) [reply osv/repos.py]: Done, added type definitions to the clone function and it's callers, also added docstrings describing each parameter.

Comment (another-rex) [reply osv/repos.py]: Can you clarify what type of input validation you think are good here? I'm not sure what I can implement for this.

**Review by oliverchang — APPROVED** [review]
thanks!

File: osv/repos.py (hunk 1 above)
Comment (andrewpollock) [reply osv/repos.py]: I think the type checking you now have will be sufficient

**Review by andrewpollock — APPROVED** [review]

**Pull request merged**